granulate_utils/java.py

```diff
@@ -215,3 +217,94 @@ def parse_jvm_version(version_string: str) -> JvmVersion:
         zing_major = None
 
     return JvmVersion(version, build, vm_name, vm_type, zing_major)
+
+
+@dataclass
+class JvmFlag:
+    name: str
+    type: str
+    value: str
+    origin: str
+    kind: List[str]
+
+    vm_flags_pattern = re.compile(
+        r"(?P<flag_type>\S+)\s+"
+        r"(?P<flag_name>\S+)\s+"
+        r"(?P<flag_equal_sign_prefix>:)?= "
+        r"(?P<flag_value>\S+)\s+"  # We don't support empty string nor spaces in flag values, although its legal values
+        r"{(?P<flag_kind>.+?)}"
+        r"(?:\s*{(?P<flag_origin_jdk_9>.*)})?"
+    )
+
+    def to_dict(self) -> Dict[str, Dict[str, Union[str, List[str]]]]:
+        return {self.name: {"type": self.type, "value": self.value, "origin": self.origin, "kind": self.kind}}
+
+    @classmethod
+    def from_dict(cls, jvm_flag_dict: Dict[str, Any]) -> JvmFlag:
+        name, flag_dict = list(jvm_flag_dict.items())[0]
+        return cls(name=name, **flag_dict)
+
+    @classmethod
+    def from_str(cls, line: str) -> Optional[JvmFlag]:
+        """
+        The output of VM.flags -all format on jdk 8:
+        bool UseCompressedClassPointers               := true                                {lp64_product}
+        flag_type flag_name                           := flag_value                          {flag_kind}
+        ":=" indicates non default origin for the flag, while "=" indicates default origin
+
+        The output of VM.flags -all format on jdk 9+:
+        bool OptoScheduling                           = false                               {C2 pd product} {default}
+        flag_type flag_name                           = flag_value                          {flag_kind} {flag_origin}
+
+        flag_kind is space separated list of kinds, e.g. "C2 pd product"
+
+        possible flag kinds:
+        "product", "manageable", "diagnostic", "experimental", "notproduct", "develop", "lp64_product", "rw", "pd", "JVMCI", "C1", "C2", "ARCH"
+        https://github.com/openjdk/jdk17u/blob/2fe42855c48c49b515b97312ce64a5a8ef3af407/src/hotspot/share/runtime/flags/jvmFlag.cpp#L338 # noqa: E501
+
+        possible flag types:
+        "bool", "int", "uint", "intx", "uintx", "uint64_t", "size_t", "double", "ccstr", "ccstrlist"
+        https://github.com/openjdk/jdk17u/blob/2fe42855c48c49b515b97312ce64a5a8ef3af407/src/hotspot/share/runtime/flags/jvmFlag.hpp#L134 # noqa: E501
+
+        possible flag origins:
+        default, non-default, command line, environment, config file, management, ergonomic, attach, internal, jimage, "command line, ergonomic" (flag is set from command line and aligned by ergonomic) # noqa: E501
```

Comment:
"command line, ergonomic" doesn't appear in that enum o.O

Reply:
https://github.com/openjdk/jdk17u/blob/2fe42855c48c49b515b97312ce64a5a8ef3af407/src/hotspot/share/runtime/flags/jvmFlag.cpp#L397

Reply:
Gotcha